ndc-sqlgen/src/main/kotlin/io/hasura/ndc/sqlgen/BaseQueryGenerator.kt

```diff
@@ -25,6 +25,108 @@ abstract class BaseQueryGenerator : BaseGenerator {
         throw NotImplementedError("Mutation not supported for this data source")
     }
 
+    protected fun findAllNativeQueries(request: QueryRequest): Set<String> {
+        val nativeQueries = mutableSetOf<String>()
+        val config = ConnectorConfiguration.Loader.config
+    
+        // Helper function to check if a collection is a native query
+        fun checkAndAddNativeQuery(collection: String) {
+            if (config.nativeQueries.containsKey(collection)) {
+                nativeQueries.add(collection)
+            }
+        }
+    
+        // Check main collection
+        checkAndAddNativeQuery(request.collection)
+    
+        // Check relationships
+        request.collection_relationships.values.forEach { rel ->
+            checkAndAddNativeQuery(rel.target_collection)
+        }
+    
+        // Recursive function to check predicates
+        fun checkPredicates(expression: Expression?) {
+            when (expression) {
+                is Expression.Exists -> {
+                    when (val collection = expression.in_collection) {
+                        is ExistsInCollection.Related -> {
+                            // Check related collection from relationship
+                            val rel = request.collection_relationships[collection.relationship]
+                                ?: error("Relationship ${collection.relationship} not found")
+                            checkAndAddNativeQuery(rel.target_collection)
+                        }
+                        is ExistsInCollection.Unrelated -> {
+                            checkAndAddNativeQuery(collection.collection)
+                        }
+                    }
+                    // Recursively check the predicate within exists
+                    checkPredicates(expression.predicate)
+                }
+                is Expression.And -> expression.expressions.forEach { checkPredicates(it) }
+                is Expression.Or -> expression.expressions.forEach { checkPredicates(it) }
+                is Expression.Not -> checkPredicates(expression.expression)
+                else -> {} // Other expression types don't reference collections
+            }
+        }
+    
+        // Check predicates in the main query
+        checkPredicates(request.query.predicate)
+    
+        // Check predicates in relationship fields
+        request.query.fields?.values?.forEach { field ->
+            if (field is IRField.RelationshipField) {
+                checkPredicates(field.query.predicate)
+            }
+        }
+    
+        return nativeQueries
+    }
+
+    fun mkNativeQueryCTEs(
```

Comment:
While the above function seems like it could well be useful across connectors, I am less sure about the one below: i.e., I am not sure that we will use this CTE strategy for all NQs. It's probably ok to leave this here for now, but it's possible that it would want to end up in specific strategies.

Reply:
Left it for now, but we can move.